In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split #We need this to split the data
import seaborn as sns

In [2]:
df = pd.read_csv("CompleteDataset.csv")
df.head()

Column1            Name  Age  \
0       47    J. Rodríguez   25   
1       22          Thiago   26   
2       31  Z. Ibrahimović   35   
3      502     L. Podolski   32   
4       55         M. Reus   28   

                                             Photo Nationality  \
0  https://cdn.sofifa.org/48/18/players/198710.png    Colombia   
1  https://cdn.sofifa.org/48/18/players/189509.png       Spain   
2   https://cdn.sofifa.org/48/18/players/41236.png      Sweden   
3  https://cdn.sofifa.org/48/18/players/150516.png     Germany   
4  https://cdn.sofifa.org/48/18/players/188350.png     Germany   

                                  Flag  Overall  Potential               Club  \
0  https://cdn.sofifa.org/flags/56.png       86         89   FC Bayern Munich   
1  https://cdn.sofifa.org/flags/45.png       88         90   FC Bayern Munich   
2  https://cdn.sofifa.org/flags/46.png       88         88  Manchester United   
3  https://cdn.sofifa.org/flags/21.png       80         80        Vissel Kobe   
4  https://cdn.sofifa.org/flags/21.png       86         86  Borussia Dortmund   

                                       Club Logo  ...    RB   RCB   RCM   RDM  \
0      https://cdn.sofifa.org/24/18/teams/21.png  ...  61.0  53.0  79.0  64.0   
1      https://cdn.sofifa.org/24/18/teams/21.png  ...  72.0  66.0  85.0  76.0   
2      https://cdn.sofifa.org/24/18/teams/11.png  ...  54.0  52.0  78.0  60.0   
3  https://cdn.sofifa.org/24/18/teams/101146.png  ...  58.0  53.0  71.0  59.0   
4      https://cdn.sofifa.org/24/18/teams/22.png  ...  63.0  52.0  80.0  64.0   

     RF    RM    RS    RW   RWB    ST  
0  83.0  82.0  80.0  83.0  65.0  80.0  
1  83.0  83.0  77.0  83.0  75.0  77.0  
2  83.0  80.0  84.0  82.0  58.0  84.0  
3  79.0  77.0  79.0  78.0  61.0  79.0  
4  86.0  84.0  82.0  86.0  67.0  82.0  

[5 rows x 75 columns]

In [3]:
#Check the data available
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17322 entries, 0 to 17321
Data columns (total 75 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Column1              17322 non-null  int64  
 1   Name                 17322 non-null  object 
 2   Age                  17322 non-null  int64  
 3   Photo                17322 non-null  object 
 4   Nationality          17322 non-null  object 
 5   Flag                 17322 non-null  object 
 6   Overall              17322 non-null  int64  
 7   Potential            17322 non-null  int64  
 8   Club                 17076 non-null  object 
 9   Club Logo            17322 non-null  object 
 10  Value                17322 non-null  object 
 11  Wage                 17322 non-null  object 
 12  Special              17322 non-null  int64  
 13  Acceleration         17322 non-null  int64  
 14  Aggression           17322 non-null  int64  
 15  Agility              17322 non-null 

### Remove the unneeded preferred positions

In [4]:
# Extract the first Preferred Position
split = df["Preferred Positions"].str.split(" ", n = 1, expand = True) 
  
# making separate column for Preferred Position
df["Preferred Position"]= split[0] 



### Subset needed data

In [5]:
# subseting the Dataframe
df_subset = df[["Preferred Position", "Acceleration","Aggression", "Agility", "Balance", "Ball control", "Composure", "Crossing", "Curve", "Dribbling", "Finishing", "Free kick accuracy", "GK diving", "GK handling", "GK kicking", "GK positioning", "GK reflexes", "Heading accuracy", "Interceptions", "Jumping", "Long passing", "Long shots", "Marking", "Penalties", "Positioning", "Reactions", "Short passing", "Shot power", "Sliding tackle", "Sprint speed", "Stamina", "Standing tackle", "Strength", "Vision", "Volleys"]]
df_subset.head()

Preferred Position  Acceleration  Aggression  Agility  Balance  \
0                 RM            77          65       82       75   
1                CDM            77          57       90       86   
2                 ST            63          84       81       41   
3                CAM            79          70       69       72   
4                 LW            86          49       86       81   

   Ball control  Composure  Crossing  Curve  Dribbling  ...  Reactions  \
0            85         87        88     87         83  ...         84   
1            92         83        72     85         90  ...         84   
2            88         91        76     82         82  ...         86   
3            78         79        75     80         78  ...         76   
4            85         83        80     90         86  ...         87   

   Short passing  Shot power  Sliding tackle  Sprint speed  Stamina  \
0             84          86              30            75       68   
1             91          75              62            68       75   
2             84          91              27            67       71   
3             77          92              39            77       68   
4             86          87              46            90       68   

   Standing tackle  Strength  Vision  Volleys  
0               42        70      86       91  
1               64        59      86       90  
2               41        88      83       90  
3               39        78      78       89  
4               36        67      86       88  

[5 rows x 35 columns]

##### Check data and run KNN

In [6]:
df_subset["Preferred Position"].value_counts()

CB     2613
ST     2189
GK     1970
CM     1898
CDM    1342
RM     1285
LM     1278
LB     1260
RB     1160
CAM    1082
RW      415
LW      411
CF      197
LWB     114
RWB     108
Name: Preferred Position, dtype: int64

In [7]:
y = df_subset["Preferred Position"]
X = df_subset[["Acceleration","Aggression", "Agility", "Balance", "Ball control", "Composure", "Crossing", "Curve", "Dribbling", "Finishing", "Free kick accuracy", "GK diving", "GK handling", "GK kicking", "GK positioning", "GK reflexes", "Heading accuracy", "Interceptions", "Jumping", "Long passing", "Long shots", "Marking", "Penalties", "Positioning", "Reactions", "Short passing", "Shot power", "Sliding tackle", "Sprint speed", "Stamina", "Standing tackle", "Strength", "Vision", "Volleys"]]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

In [8]:
from sklearn.neighbors import KNeighborsClassifier #the object class we need
knn = KNeighborsClassifier(n_neighbors=5) #create a KNN-classifier with 5 neighbors (default)
knn = knn.fit(X_train, y_train) #this fits the k-nearest neigbor model with the train data
knn.score(X_test, y_test) #calculate the fit on the test data

0.4970175101019819

##### Our overall KNN score is .4970 so we are correctly predicting position in ~50% of cases. Now lets look deeper at the precision and recal per position

In [9]:
from sklearn.metrics import classification_report
y_pred = knn.predict(X_test) 
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         CAM       0.19      0.26      0.22       306
          CB       0.67      0.80      0.73       760
         CDM       0.29      0.32      0.31       408
          CF       0.31      0.07      0.12        67
          CM       0.39      0.42      0.41       580
          GK       1.00      1.00      1.00       607
          LB       0.34      0.38      0.36       372
          LM       0.21      0.18      0.20       381
          LW       0.18      0.09      0.12       121
         LWB       0.00      0.00      0.00        35
          RB       0.31      0.24      0.27       362
          RM       0.22      0.17      0.19       378
          RW       0.26      0.08      0.12       125
         RWB       0.00      0.00      0.00        34
          ST       0.70      0.80      0.74       661

    accuracy                           0.50      5197
   macro avg       0.34      0.32      0.32      5197
weighted avg       0.47   

#### We can clearly see that the algorithim is better at picking some positions more than others. 

For example Goalkeepers were flawlessly predicted by the model. This is unsurprising when you consider that Goalkeepers has significantly higher stats than outfield players in relation to GK values. 

Additionally, Strikers were predicted at a relativly high rate. This is again likely due to the high reliance the position has on doing one job, scoring goals. This is very different for example in comparison to a midfielder, who could be a defensive player, attacking second striker, or a creative player. Couple this with the fact that these players are split across CF, CAM, CM and CDM. 

#### Maybe its better to use a random forrest algorithim instead?

In [10]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(random_state=1, n_estimators=100) #RF is a random algorithm, so to get the same results we need to use random_state
rf = rf.fit(X_train, y_train)

In [27]:
y_pred = rf.predict(X_test) #the predicted values
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         CAM       0.29      0.25      0.27       306
          CB       0.69      0.85      0.76       760
         CDM       0.36      0.25      0.30       408
          CF       0.90      0.13      0.23        67
          CM       0.41      0.59      0.48       580
          GK       1.00      1.00      1.00       607
          LB       0.37      0.52      0.43       372
          LM       0.25      0.19      0.22       381
          LW       0.25      0.02      0.05       121
         LWB       1.00      0.03      0.06        35
          RB       0.34      0.20      0.26       362
          RM       0.26      0.22      0.24       378
          RW       0.31      0.03      0.06       125
         RWB       0.00      0.00      0.00        34
          ST       0.66      0.87      0.75       661

    accuracy                           0.54      5197
   macro avg       0.47      0.34      0.34      5197
weighted avg       0.51   

/opt/anaconda3/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#### The random forrest results are marginally better than the KNN output. An much higher average precision and marginally higher recall.

However the change is minimal, while it's better in some ways (the LWB prediction got all LWB players!) its worse in others (it only got 3% of them...).